src/core/qgsvectorlayereditbuffer.cpp

```diff
@@ -310,6 +310,36 @@ bool QgsVectorLayerEditBuffer::commitChanges( QStringList &commitErrors )
   // no                 yes                   => changeAttributeValues
   // yes                yes                   => changeFeatures
 
+  // to fix https://issues.qgis.org/issues/15741
+  // first of all check if feature to add is compatible with provider type
+  // this check have to be done before all checks to avoid to clear internal
+  // buffer if some of next steps success.
+  if ( success && !mAddedFeatures.isEmpty() )
+  {
+    if ( cap & QgsVectorDataProvider::AddFeatures )
+    {
+      for ( QgsFeature f : mAddedFeatures )
+      {
+        if ( ( ! f.hasGeometry() ) ||
+             ( f.geometry().wkbType() == provider->wkbType() ) )
```

Comment:
I would like to agree, but convertToProviderType reuturn nullptr in case conversion is not necessary as in case conversion is not possible, so I can't distinguish among a normal workflow (skip geometry) from an error.
That's the reason of this check (that is redundant inside convertToProviderType).
I would refactor convertToProviderType to return the same geom instead of nullptr, btw we would have the problem in case geometry is not present.